### ISSUE TITLE
og_image per post

### ISSUE THREAD

**Issue by juliohm** [mmistakes/minimal-mistakes#943]
Is it possible to assign an og_image on a per post basis? I have a data visualization as a SVG figure, but it is not captured by the theme when I share the link on LinkedIn for instance: https://juliohm.github.io/science/DAG-of-Julia-packages/

**Comment by mmistakes** [comment]
Yes. You need to use a page header image.

```yaml
header:
  image: path-to-image.jpg
```

https://mmistakes.github.io/minimal-mistakes/docs/layouts/#headers

**Comment by juliohm** [comment]
I know it works with headers, I wonder if we can get the same effect without a header?

**Comment by mmistakes** [comment]
Not currently. You could hack the theme to do it though. 

https://github.com/mmistakes/minimal-mistakes/blob/master/_includes/seo.html#L63-L73

**Comment by juliohm** [comment]
Awesome @mmistakes , I think I can hack it to achieve what I want, feel free to close the issue if you feel this is not a general feature request that would be useful to other users.

**Comment by mmistakes** [comment]
Something like this should do the trick:

```html
  {% if page.header.image %}
    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:image" content="{% if page.header.image contains "://" %}{{ page.header.image }}{% else %}{{ page.header.image | absolute_url }}{% endif %}">
  {% else %}
    <meta name="twitter:card" content="summary">
    {% if page.header.teaser %}
      <meta name="twitter:image" content="{% if page.header.teaser contains "://" %}{{ page.header.teaser }}{% else %}{{ page.header.teaser | absolute_url }}{% endif %}">
    {% elsif page.og_image %}
      <meta name="twitter:image" content="{{ page.og_image | absolute_url }}">
    {% elsif site.og_image %}
      <meta name="twitter:image" content="{{ site.og_image | absolute_url }}">
    {% endif %}
  {% endif %}
```

**Comment by juliohm** [comment]
@mmistakes can you please comment the difference between header.teaser and the proposed page.og_image, couldn't they serve the same purpose?

**Comment by mmistakes** [comment]
They could. Guess it depends on how you expect to use them. `page.header.teaser` are optional smaller images used in grid view lists (e.g. related posts module).

![related_posts](https://cloud.githubusercontent.com/assets/1376749/24917678/e9255ba6-1eab-11e7-82ae-043bc7c3bd18.jpg)

So if you populate that field in your post/page YAML the image will be used for OpenGraph and will also appear on the page in certain locations. 

If you don't want that behavior then the additional of `page.og_image` to _includes/seo.html would fit the bill.

**Comment by juliohm** [comment]
I was just wondered if they could be merged into a single option. But as you said, maybe finer granularity is desired.

**Comment by mmistakes** [comment]
Yes they could.

In general there's a lot of refactoring and cleanup that could be done to the `/_includes/seo.html`. I plan on replacing it with [Jekyll's SEO tag plugin](https://github.com/mmistakes/minimal-mistakes/issues/324), but haven't gotten around to it yet.

But essentially something like this that sets defaults could work. The logic being... if `page.header.image` is defined use that otherwise fallback to `page.header.teaser` which fallsback to `page.og_image`.

```
{% assign seo_page_image = page.header.image | default: page.header.teaser | default: page.og_image %}
{% unless seo_page_image contains "://" %}
  {% assign seo_page_image = seo_page_image | absolute_url %}
{% endif %}

{% if seo_page_image %}
  <meta property="og:image" content="{{ seo_page_image }}">
{% endif %}

**Comment by juliohm** [comment]
@mmistakes I tried adding the lines

```julia
{% elsif page.og_image %}
  <meta property="og:image" content="{% if page.og_image contains "://" %}{{ page.og_image }}{% else %}{{ page.og_image | absolute_url }}{% endif %}">
```

to `seo.html`, then adding the line `og_image: /assets/images/some_image.png` to the YAML header of a post. The result doesn't affect the preview image when I copy/paste the link on LinkedIn: https://juliohm.github.io/science/DAG-of-Julia-packages/

Could you please help with the issue?

**Comment by mmistakes** [comment]
The [OpenGraph Debugger](https://developers.facebook.com/tools/debug/) is your friend. The `og:image` is correct in the HTML so that part is right. OpenGraph data is cached (not sure for how long) which is why it's showing the old image.

![og_image](https://cloud.githubusercontent.com/assets/1376749/25004074/799bd5d2-2020-11e7-8d2d-04bb691494f7.jpg)

Hit rescrape to force it to refresh.

![scraped](https://cloud.githubusercontent.com/assets/1376749/25004103/99899d70-2020-11e7-907f-06257420afd4.jpg)

**Comment by juliohm** [comment]
Your answers are always so good! Thank you for the clarification.

How long usually it takes for this cache to be updated and shared links to have the correct image?

**Comment by mmistakes** [comment]
No problem.

Not entirely sure on the caching. I know as far as Facebook is concerned when I've tried to share a link and noticed the image was wrong, if I rescraped it with the debug tool then tried to share again it would be updated.

But I don't know if it's immediate for links that have already been shared. That's a question for Facebook and LinkedIn.

Going to close this as you seem to have it working.
